### PR TITLE
Add Request-Attempt header to CAPI requests

### PR DESCRIPTION
**Pull request by JustinPinner**
When requests are made through the backoff object, we keep a counter of the number of attempts we've made for the resource.

To help us understand how the backoff implementations are working it'd be really handy to be able to see how frequently the retry code is activating, and we'd like to see that in our server logs, so we figured adding a header to every request was a reasonable way of achieving that.

While it's tempting to _not_ send a header _unless_ it's a retry attempt, sending it with a zero value for the initial request will be useful as we'll be able to know for certain that clients have started using the new library version as we expect most of the requests to have a Request-Attempt value of 0 (zero) - and the version number reported from the client has also been problematic so only when we know that we can rely on that will we remove the initial (zero value) header.

There is a corresponding PR in our concierge service to use log out the new header value for examination during periods of service instability.

### CODE REVIEW

File: client/src/main/scala/com.gu.contentapi.client/ContentApiClient.scala [hunk 1]
```diff
@@ -41,7 +41,8 @@ trait ContentApiClient {
   def get(url: String, headers: Map[String, String])(implicit context: ExecutionContext): Future[HttpResponse]
 
   /** Some HTTP headers sent along each CAPI request */
-  private def headers = Map("User-Agent" -> userAgent, "Accept" -> "application/x-thrift")
+  private def headers(retry: Int) =
+    Map("User-Agent" -> userAgent, "Accept" -> "application/x-thrift", "x-retry" -> s"$retry")
```
Comment (regiskuckaertz) [inline client/src/main/scala/com.gu.contentapi.client/ContentApiClient.scala]: Since a lot of those will be 0, can we make it an `Option[Int]` instead, and only add the header when there were actual retries?

File: client-default/src/test/scala/com.gu.contentapi.client/GuardianContentClientTest.scala [hunk 2]
```diff
@@ -71,11 +71,12 @@ class GuardianContentClientTest extends FlatSpec with Matchers with ScalaFutures
     content.futureValue.id should be (TestItemPath)
   }
 
-  it should "perform a given removed content query" in {
```
Comment (anshulbajpai) [inline client-default/src/test/scala/com.gu.contentapi.client/GuardianContentClientTest.scala]: This seems to be an unrelated change. Was this an irrelevant/unuseful test?

Comment (JustinPinner) [reply client-default/src/test/scala/com.gu.contentapi.client/GuardianContentClientTest.scala]: That's related to the `I've temporarily made that just check for a 404 to move this along` comment in the PR body :) It'll disappear once I've merged from master now that the 15.6 CAPI models changes are in there.

File: client/src/main/scala/com.gu.contentapi.client/ContentApiClient.scala (hunk 1 above)
Comment (JustinPinner) [reply client/src/main/scala/com.gu.contentapi.client/ContentApiClient.scala]: I'd rather see the zero to begin with to know that it's working. It's currently not coming through anywhere at all.

File: client/src/main/scala/com.gu.contentapi.client/ContentApiBackoff.scala [hunk 3]
```diff
@@ -30,10 +30,10 @@ abstract class ContentApiBackoff(implicit executor: ScheduledExecutor) extends P
 
   def currentState: ContentApiBackoff = self
 
-  def execute(operation: => Future[HttpResponse])(implicit context: ExecutionContext): Future[HttpResponse] = {
+  def execute(operation: Int => Future[HttpResponse])(implicit context: ExecutionContext): Future[HttpResponse] = {
```
Comment (Mullefa) [inline client/src/main/scala/com.gu.contentapi.client/ContentApiBackoff.scala]: This breaking API change is associated with a minor upgrade? Is this ok? i.e. do we not adhere to semver?

Comment (JustinPinner) [reply client/src/main/scala/com.gu.contentapi.client/ContentApiBackoff.scala]: I think execute is not directly called from outside - I'll re-check. But actually, maybe a major version bump is the way to go for this anyway.

Comment (JustinPinner) [reply client/src/main/scala/com.gu.contentapi.client/ContentApiBackoff.scala]: So `backoffStrategy.execute` is called from within `private def fetchResponse` so I think we're protecting consumers of the library from that change. I'll bump to 16.0 anyway though I think.
